cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/TaskWorkerHttpHandlerInternal.java

```diff
@@ -169,31 +166,49 @@ private void enablePeriodicRestart(CConfiguration cConf,
         Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION_SECOND, 0);
     int lowerBound = (int) (duration - duration * DURATION_FRACTION);
     int upperBound = (int) (duration + duration * DURATION_FRACTION);
-    if (lowerBound > 0) {
-      int waitTime =
-          (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
-      Executors.newSingleThreadScheduledExecutor(
-              Threads.createDaemonThreadFactory("task-worker-restart"))
-          .scheduleWithFixedDelay(
-              () -> {
-                if (mustRestart.get()) {
-                  // We force pod restart as the ongoing request has not finished since last
-                  // periodic restart check.
-                  stopper.accept("");
-                  return;
-                }
-                // we restart once ongoing request (which has set runningRequestCount to 1)
-                // finishes.
-                mustRestart.set(true);
-                if (runningRequestCount.compareAndSet(0, 1)) {
-                  // there is no ongoing request. pod gets restarted.
-                  stopper.accept("");
-                }
-              },
-              waitTime,
-              waitTime,
-              TimeUnit.SECONDS);
+
+    if (lowerBound <= 0) {
+      return;
+    }
+    int waitTime = (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
+
+    int taskDeadlineSeconds = cConf.getInt(
```

Comment:
nit: extract this + next if as a method. This way you won't need this `taskDeadline/finalTaskDeadline` pair

Reply:
Done